File: Merge Sort/mergeSort.py
'''
    @author: Guilherme Vieira Silva Gonçalves
    @course: Ciência da Computação, 5th semester
'''

def merge(v, p, q, r):
    n1 = q - p + 1
    n2 = r - q
    E = [0] * (n1 + 1)
    D = [0] * (n2 + 1)
    for i in range(n1):
        E[i] = v[p + i]
    for j in range(n2):
        D[j] = v[q + j + 1]

    E[n1] = float('inf')
    D[n2] = float('inf')
    i = j = 0
    for k in range(p, r + 1):
        if E[i] <= D[j]:
            v[k] = E[i]
            i += 1
        else:
            v[k] = D[j]
            j += 1


def mergeSort(v, p, r):
    if p < r:
        q = (p + r) // 2
        mergeSort(v, p, q)
        mergeSort(v, q + 1, r)
        merge(v, p, q, r)

def main():
    arr = [10, 7, 3, 8, 4, 2, 5]
    n = len(arr)
    mergeSort(arr, 0, n - 1)
    print("Array ordenado: ", end="")
    for i in range(n):
        print(arr[i], end=" ")
    print()

if __name__ == "__main__":
    main()

File: Merge Sort/mergeSort.cpp
#include <iostream>
#include <climits>

void merge(int v[], int p, int q, int r){
    int n1 = q - p + 1;
    int n2 = r - q;
    int E[n1 + 1];
    int D[n2 + 1];
    for (int i = 0; i < n1; i++) {
        E[i] = v[p + i];

    }
    for (int j = 0; j < n2; j++) {
        D[j] = v[q + j + 1];
    }
    E[n1] = INT_MAX;
    D[n2] = INT_MAX;
    int i = 0;
    int j = 0;
    for (int k = p; k <= r; k++) {
        if (E[i] <= D[j]) {
            v[k] = E[i];
            i++;
        }
        else{
            v[k] = D[j];
            j++;
        }
    }

}

void mergeSort(int v[], int p, int r){
    if (p < r) {
        int q = (p + r)/2;
        mergeSort(v, p, q);
        mergeSort(v, q + 1, r);
        merge(v, p, q, r);
    }
}

int main() {
    int arr[] = {10, 7, 3, 8, 4, 2, 5};
    int n = sizeof(arr) / sizeof(arr[0]);
    mergeSort(arr, 0, n - 1);
    for (int i = 0; i < n; i++) {
        std::cout << arr[i] << " ";
    }
    std::cout << std::endl;
}
